operator/internal/controller/pulumi/git.go

```diff
@@ -18,318 +18,276 @@ import (
-				Password: repoArgs.Auth.PersonalAccessToken,
-			}
+	if gs.auth.SSHPrivateKey != "" {
+		return ssh.NewPublicKeys("git", []byte(gs.auth.SSHPrivateKey), gs.auth.Password)
+	}
+
+	return nil, nil
+}
+
+// resolveGitAuth sets up the authentication option to use for the git source
+// repository of the stack. If neither gitAuth or gitAuthSecret are set,
+// a pointer to a zero value of GitAuth is returned — representing
+// unauthenticated git access.
+//
+// The workspace pod is also mutated to mount these references at some
+// well-known paths.
+func (sess *StackReconcilerSession) resolveGitAuth(ctx context.Context) (*auto.GitAuth, error) {
+	auth := &auto.GitAuth{}
+
+	if sess.stack.GitSource == nil {
+		return auth, nil // No git source to auth.
+	}
+
+	if sess.stack.GitAuthSecret == "" && sess.stack.GitAuth == nil {
+		return auth, nil // No authentication.
+	}
+
+	if sess.stack.GitAuthSecret != "" {
+		namespacedName := types.NamespacedName{Name: sess.stack.GitAuthSecret, Namespace: sess.namespace}
+
+		// Fetch the named secret.
+		secret := &corev1.Secret{}
+		if err := sess.kubeClient.Get(ctx, namespacedName, secret); err != nil {
+			sess.logger.Error(err, "Could not find secret for access to the git repository",
+				"Namespace", sess.namespace, "Stack.GitAuthSecret", sess.stack.GitAuthSecret)
+			return nil, err
 		}
 
-		// then we check to see if a username and a password has been specified
-		if authDetails.Password != "" && authDetails.Username != "" {
-			cloneOptions.Auth = &http.BasicAuth{
-				Username: repoArgs.Auth.Username,
-				Password: repoArgs.Auth.Password,
+		// First check if an SSH private key has been specified.
+		if sshPrivateKey, exists := secret.Data["sshPrivateKey"]; exists {
+			auth.SSHPrivateKey = string(sshPrivateKey)
+
+			if password, exists := secret.Data["password"]; exists {
+				auth.Password = string(password)
+			}
+			// Then check if a personal access token has been specified.
+		} else if accessToken, exists := secret.Data["accessToken"]; exists {
+			auth.PersonalAccessToken = string(accessToken)
+			// Then check if basic authentication has been specified.
+		} else if username, exists := secret.Data["username"]; exists {
+			if password, exists := secret.Data["password"]; exists {
+				auth.Username = string(username)
+				auth.Password = string(password)
+			} else {
+				return nil, fmt.Errorf(`creating gitAuth: no key "password" found in secret %s`, namespacedName)
 			}
 		}
+
+		return auth, nil
 	}
 
-	// *Repository.Clone() will do appropriate fetching given a branch name. We must deal with
-	// different varieties, since people have been advised to use these as a workaround while only
-	// "refs/heads/<default>" worked.
-	//
-	// If a reference name is not supplied, then .Clone will fetch all refs (and all objects
-	// referenced by those), and checking out a commit later will work as expected.
-	if repoArgs.Branch != "" {
-		refName := plumbing.ReferenceName(repoArgs.Branch)
```

Comment:
There's some sophistication here in decoding the reference.  Should some of this logic survive?

Reply:
Not worth keeping -- this is more to do with the actual repo checkout, which we're now offloading to the automation API. For example `refs/remotes/origin` only exists after checkout, so the `IsRemote` logic isn't relevant. The `IsBranch`/`IsTag` checks are handled on line 112 now:
```
if r.Name().String() == gs.ref || r.Name().Short() == gs.ref
```
That is, we take commits belonging to an exact-matching ref (e.g. `refs/heads/master`) or the first matching short ref (e.g. `master`).